File: api_gateway_service/infra/outbox/pruning.go
package outbox

import (
	"context"
	"database/sql"
	"log/slog"
	"os"
	"strconv"
	"time"
)

// PruneRunner periodically deletes old processed outbox rows.
type PruneRunner struct {
	db            *sql.DB
	retentionDays int
	interval      time.Duration
	cancel        context.CancelFunc
	done          chan struct{}
}

func NewPruneRunner(db *sql.DB) *PruneRunner {
	days := 30
	if v := os.Getenv("PRUNE_RETENTION_DAYS"); v != "" {
		if d, err := strconv.Atoi(v); err == nil && d > 0 {
			days = d
		}
	}

	hours := 24
	if v := os.Getenv("PRUNE_INTERVAL_HOURS"); v != "" {
		if h, err := strconv.Atoi(v); err == nil && h > 0 {
			hours = h
		}
	}

	return &PruneRunner{
		db:            db,
		retentionDays: days,
		interval:      time.Duration(hours) * time.Hour,
		done:          make(chan struct{}),
	}
}

func (p *PruneRunner) Start() {
	ctx, cancel := context.WithCancel(context.Background())
	p.cancel = cancel

	go func() {
		defer close(p.done)
		slog.Info("outbox pruning started", "retentionDays", p.retentionDays, "intervalHours", p.interval.Hours())

		// Run immediately, then on interval
		p.prune(ctx)

		ticker := time.NewTicker(p.interval)
		defer ticker.Stop()

		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				p.prune(ctx)
			}
		}
	}()
}

func (p *PruneRunner) Stop() {
	if p.cancel != nil {
		p.cancel()
		<-p.done
		slog.Info("outbox pruning stopped")
	}
}

func (p *PruneRunner) prune(ctx context.Context) {
	cutoff := time.Now().Add(-time.Duration(p.retentionDays) * 24 * time.Hour)
	query := `DELETE FROM outbox WHERE processed_at IS NOT NULL AND processed_at < $1`

	result, err := p.db.ExecContext(ctx, query, cutoff)
	if err != nil {
		slog.Error("outbox pruning failed", "error", err)
		return
	}

	rows, _ := result.RowsAffected()
	slog.Info("outbox pruning complete", "rowsDeleted", rows)
}

File: hotel_service/src/infra/outbox/pruning.ts
/**
 * Outbox table pruning for hotel-service.
 * Deletes processed outbox rows older than PRUNE_RETENTION_DAYS.
 */

import { prisma } from '../database/prisma.js';
import { logger } from '../logger/index.js';

const RETENTION_DAYS = Number(process.env.PRUNE_RETENTION_DAYS) || 30;
const INTERVAL_MS =
    Number(process.env.PRUNE_INTERVAL_HOURS || 24) * 60 * 60 * 1000;

let timer: ReturnType<typeof setInterval> | null = null;

async function pruneOutbox(): Promise<number> {
    const cutoff = new Date(Date.now() - RETENTION_DAYS * 86_400_000);
    const result = await prisma.outbox.deleteMany({
        where: {
            processedAt: { not: null, lt: cutoff },
        },
    });
    return result.count;
}

async function runPruning(): Promise<void> {
    try {
        const count = await pruneOutbox();
        logger.info('Outbox pruning complete', { outboxRowsDeleted: count });
    } catch (err) {
        logger.error('Outbox pruning job failed', { error: err });
    }
}

export function startOutboxPruning(): void {
    logger.info(
        `Starting outbox pruning — retention=${RETENTION_DAYS}d, interval=${INTERVAL_MS / 3_600_000}h`
    );
    runPruning();
    timer = setInterval(runPruning, INTERVAL_MS);
}

export function stopOutboxPruning(): void {
    if (timer) {
        clearInterval(timer);
        timer = null;
        logger.info('Outbox pruning stopped');
    }
}
